Support time lag retrieval for subscription stats

mtesseract opened the pull request:
This fixes the per-subscription statistics retrieval and implements support for the new `show_time_lag` parameter of the Nakadi API.

The statistics retrieval never worked before, thus I felt free to improve the general API in this regard.

In particular I have renamed the type `SubscriptionEventTypeStatsResult` to `SubscriptionStats` — the Nakadi Swagger does not contain a specific name for this object type.

There is some reformatting being done by brittany, sorry for the noise.
I have tested the API call manually, but a test should be added.

I wouldn't mind merging this right away, though.

etorreborre commented on src/Network/Nakadi/Subscriptions/Stats.hs:
```diff
@@ -18,39 +18,52 @@ API.
-  <> "/cursors"
+  "/subscriptions/" <> subscriptionIdToByteString subscriptionId <> "/stats"
 
 -- | @GET@ to @\/subscriptions\/SUBSCRIPTION\/cursors@. Low level
```
This comment needs to be updated with `/stats` instead of `/cursors`. Same thing for the other function below.

mtesseract replied:
Done

etorreborre commented on src/Network/Nakadi/Subscriptions/Stats.hs:
```diff
@@ -18,39 +18,52 @@ API.
+  :: MonadNakadi b m
   => SubscriptionId                        -- ^ Subscription ID
+  -> [SubscriptionStatsParameter]
   -> m (Map EventTypeName [PartitionStat]) -- ^ Subscription
```
I am generally not satisfied with returning `Map` as a datatype from an API because it feels like exposing implementation details. `SubscriptionStats` seems to be fine because it maps 1-to-1 with the JSON returned by Nakadi but maybe I would reserve this name for a wrapper around the `Map` and would find another name for the "low-level".

mtesseract replied:
I'm not sure I understand the criticism regarding `Map`. But maybe we can discuss possible API changes related to this question in a separate issue / PR and restrict this one to implement the show_time_lag functionality?

What about this approach of providing parameter flags as a list? Does this seem reasonable?

Thanks for your review.

etorreborre replied:
My experience with providing maps in API hasn't been so good but possibly because the underlying domain was changing too much, using different keys or having additional keyed values. In the case of Nakadi I suppose the API is kind of stable or not so changing that we need to encapsulate every map.

I guess the drawbacks with parameters as a list is that you make mistakes by having more than one in the list. Maybe one alternative, which would work because there aren't too many of them, would be to use the typelevel-sets project?  If there are too many parameters the compiler becomes dog-slow :-(.

mtesseract replied:
I also assume that we can safely assume that the return value of `subscriptionStats` yields a map keyed by `EventTypeName` and providing (possibly many) per-partition statistics. If this changes, I guess we are talking about a different / new endpoint.
Translating the value from some opaque `SubscriptionStats` to some more concrete `Map` removes — in my opinion — unnecessary unwrapping / encapsulattion.

Not sure if I correctly understand your argument regarding parameters as a list. I *think* you are referring to the situation of mutually exclusive parameters and the possibility to reject programs which are invalid in this regard, is this correct? i.e., nobody would prevent me from providing `[ShowTimeLag, WithoutTimeLag]`? Personally I have no experience with type-level-sets. But before going type level I would find it very important to have a clear picture of the pros/cons situation.

Don't get me wrong, I am really not sure if this parameters-as-list approach is the most reasonably style to implement here. Especially since we should find a way that does not only satisfy the requirements for a single endpoint, but might be a more general approach for configuring specific Nakadi requests.

Alternatively, we could also change the function such that it reads:
```
subscriptionStats :: SubscriptionStatsConf -> SubscriptionId -> ...
```

where the constructors for `SubscriptionStatsConf` are not exported, instead of export some smart constructors:

```
subscriptionStatsConfWithTimeLag :: SubscriptionStatsConf
subscriptionStatsConfPlain :: SubscriptionStatsConf
```

In terms of extendability, this would allow us to provide additional smart constructors in the future, if the API allows even more configurability.

mtesseract replied:
Or we just take the easy way out: We already have a configuration object, it is implicitly contained in the Nakadi monad.

How about we simply implement:

```
setShowTimeLag :: Bool -> Config m -> Config m
```
Then, subscriptionStats can remain simple:

```
subscriptionStats :: MonadNakadi b m => SubscriptionId -> m (Map EventTypeName [PartitionStat])
```

And users can also it as follows:

```
local (setShowTimeLag %~ True) $ subscriptionStats ...
```

I think I would prefer this way!

etorreborre replied:
That seems reasonable indeed :-)

etorreborre commented:
👍

mtesseract commented on tests/Network/Nakadi/Subscriptions/Test.hs:
```diff
@@ -12,81 +12,85 @@ import           ClassyPrelude
   ]
 
 testSubscriptionsList :: Config App -> Assertion
-testSubscriptionsList conf = runApp . runNakadiT conf $ do
```
I did this change in order to reduce formatting noise by brittany. Otherwise the whole function would have been reindented.

mtesseract commented on src/Network/Nakadi/Config.hs:
```diff
@@ -60,20 +64,21 @@ newConfig
   => HttpBackend b
   -> Request           -- ^ Request Template
   -> Config b          -- ^ Resulting Configuration
-newConfig httpBackend request =
```
Formatting noise, only `_subscriptionStats` has been added.

mtesseract commented on src/Network/Nakadi/Config.hs:
```diff
@@ -84,8 +89,8 @@ newConfigIO = newConfig httpBackendIO
 
 -- | Produce a new configuration, with optional HTTP manager settings
 -- and mandatory HTTP request template.
```
Most of this is formatting noise. Only `setShowTimeLag` is new.

mtesseract commented on src/Network/Nakadi/Internal/Util.hs:
```diff
@@ -18,38 +18,39 @@ module Network.Nakadi.Internal.Util
   , decodeThrow
   , sequenceSnd
   , extractQueryParametersFromPath
```
With the exception of `encodeStrict` everything is formatting noise here.

mtesseract commented on tests/Network/Nakadi/Subscriptions/Test.hs:
```diff
@@ -12,81 +12,85 @@ import           ClassyPrelude
+import           Network.Nakadi.Subscriptions.Processing.Test
+import           Network.Nakadi.Subscriptions.Stats.Test
 
 testSubscriptions :: Config App -> TestTree
```
More noise, I have simply added the new tests in this module.

mtesseract commented:
As discussed, I have added the `setShowTimeLag` function and simplified the `subscriptionStats` API.

When I compared our model for `PartitionStat` with the Nakadi API (https://nakadi.io/manual.html#definition_SubscriptionEventTypeStats), I have noticed that we wrongly regarded `stream_id` and `unconsumed_events` as mandatory. I have fixed this.

I am currently trying out the Haskell IDE Engine (HIE), it comes bundled with brittany integration (https://github.com/lspitzner/brittany), which has sadly added quite some formatting noise, sorry for that.

I have also added missing tests for subscriptionStats.

Ok to merge?

etorreborre commented on tests/Network/Nakadi/Subscriptions/Stats/Test.hs:
```diff
@@ -0,0 +1,84 @@
+          & concat
+          & map (^. L.consumerLagSeconds)
+          & filter isNothing
+  liftIO $ [] @=? withoutTimeLags
```
Can we also show that there is at least one time lag in the list, because it might as well be empty?

etorreborre commented:
👍 ok to merge but I think that one test could be improved.

etorreborre commented:
👍
